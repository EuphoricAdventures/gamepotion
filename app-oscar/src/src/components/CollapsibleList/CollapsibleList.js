import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import Button from '../Button/Button'
import List from '../List/List'

const StyledCollapsibleList = styled.ul`
  .component--button {
    width: 100%;
    opacity: 0.5;
  }
  .component--button + .component--list {
    margin-top: 0.5rem;
  }
`

class CollapsibleList extends React.PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      isCollapsed: false
    }
    this.onUpdateCollapsed = this.onUpdateCollapsed.bind(this)
  }

  onUpdateCollapsed() {
    this.setState({
      isCollapsed: !this.state.isCollapsed
    })
  }

  render() {
    return (
      <StyledCollapsibleList className='component--list'>
        <Button flavour='weak' onClick={this.onUpdateCollapsed}>{this.state.isCollapsed ? 'Show' : 'Hide'} {this.props.label}</Button>
        {this.state.isCollapsed === false &&
          <List>
            {this.props.children}
          </List>
        }
      </StyledCollapsibleList>
    )
  }
}

CollapsibleList.propTypes = {
  label: PropTypes.string.isRequired,
  isCollapsed: PropTypes.bool
}

CollapsibleList.defaultProps = {
  isCollapsed: false
}

export default CollapsibleList
