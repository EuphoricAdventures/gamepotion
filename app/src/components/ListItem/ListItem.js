import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import classnames from 'classnames'

import icons from '../../icons'
import { font, colours } from '../../styleAbstractions'

const StyledListItem = styled.button`
  position: relative;
  display: block;
  width: 100%;
  height: calc(2rem + 4px);
  list-style-type: none;
  border-width: 2px;
  border-style: solid;
  border-color: transparent;
  border-radius: 4px;
  background-color: transparent;
  transition: border-color 0.2s ease-in-out;
  outline: 0;
  &.selected {
    background-color: #dadfe1;
  }
  &:focus {
    border-color: ${colours.outline};
  }
  > img {
    display: block;
    float: left;
    width: 1.5rem;
    height: 1.5rem;
    margin-top: 0.25rem;
    margin-left: 0.25rem;
    // background-color: red;
  }
  > span {
    display: block;
    float: left;
    height: 2rem;
    line-height: 2rem;
    margin-left: 0.5rem;
    ${font}
    color: ${colours.fore};
    // background-color: green;
  }
  &:hover {
    > .actions {
      visibility: visible;
      opacity: 1;
    }
  }
  > .actions {
    position: absolute;
    top: 0;
    right: 0;
    height: 2rem;
    transition: opacity 0.1s ease-in-out;
    visibility: hidden;
    opacity: 0;
    // background-color: pink;
    img {
      display: block;
      float: left;
      width: 1.5rem;
      height: 1.5rem;
      margin-top: 0.25rem;
      margin-left: 0.25rem;
      background-color: transparent;
      opacity: 0.5;
      border-radius: 4px;
      transition: opacity 0.1s ease-in-out;
      outline: 0;
      // background-color: blue;
      &:focus {
        opacity: 1;
        background-color: ${colours.outline};
      }
      &:hover {
        opacity: 1;
      }
    }
  }
`

const actions = {
  'add': (id, handleOnAction) => (<img title='Add' key='action-add' onClick={(e) => handleOnAction(e, id, 'add')} className='action' src={icons.generic.actions.add} alt='' tabIndex='0' />),
  'rename': (id, handleOnAction) => (<img title='Rename' key='action-rename' onClick={(e) => handleOnAction(e, id, 'rename')} className='action' src={icons.generic.actions.edit} alt='' tabIndex='0' />),
  'delete': (id, handleOnAction) => (<img title='Delete' key='action-delete' onClick={(e) => handleOnAction(e, id, 'delete')} className='action' src={icons.generic.actions.delete} alt='' tabIndex='0' />)
}

const getAction = (onAction, id, name) => {
  const foundAction = actions[name]
  if (typeof foundAction === 'function') {
    return foundAction(id, onAction)
  }
  return null
}

const ListItem = ({ id, icon, selected, actions, children, onChoose, onAction }) => {
  const handleOnAction = (e, id, action) => {
    e.stopPropagation()
    onAction(id, action)
  }
  return (
    <StyledListItem onClick={() => onChoose(id)} className={classnames('component--list-item', {'selected': selected})}>
      <img src={icon} alt='' />
      <span>{children}</span>
      <div className='actions'>
        {actions.map(name => getAction(handleOnAction, id, name))}
      </div>
    </StyledListItem>
  )
}

ListItem.propTypes = {
  id: PropTypes.string.isRequired,
  icon: PropTypes.string,
  selected: PropTypes.bool,
  actions: PropTypes.array,
  onChoose: PropTypes.func,
  onAction: PropTypes.func
}

ListItem.defaultProps = {
  icon: '',
  selected: false,
  actions: [],
  onChoose: () => {},
  onAction: () => {}
}

export default ListItem
