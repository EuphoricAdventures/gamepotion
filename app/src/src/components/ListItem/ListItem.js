import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import classnames from 'classnames'

import icons from '../../icons'

const styleAbstractions = {
  colours: {
    'outline': '#a537fd',
    'fore': '#2e3131',
  },
  font: `
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", sans-serif;
    font-weight: 400;
    text-size-adjust: 100%;
    -webkit-font-smoothing: antialiased;
  `
}

const StyledListItem = styled.li`
  position: relative;
  display: block;
  width: calc(100% - 4px);
  height: 2rem;
  list-style-type: none;
  border-width: 2px;
  border-style: solid;
  border-color: transparent;
  border-radius: 4px;
  background-color: transparent;
  transition: border-color 0.2s ease-in-out;
  outline: 0;
  cursor: default;
  &.selected {
    background-color: #dadfe1;
  }
  &:focus {
    border-color: ${styleAbstractions.colours.outline};
  }
  > img {
    display: block;
    float: left;
    width: 1.5rem;
    height: 1.5rem;
    margin-top: 0.3rem;
    margin-bottom: 0.3rem;
    margin-left: 0.3rem;
    // background-color: red;
  }
  > span {
    display: block;
    float: left;
    height: 1.5rem;
    line-height: 1.5rem;
    margin-top: 0.3rem;
    margin-bottom: 0.3rem;
    margin-left: 0.5rem;
    ${styleAbstractions.font}
    color: ${styleAbstractions.colours.fore};
    // background-color: green;
  }
  &:hover {
    > .actions {
      visibility: visible;
      opacity: 1;
    }
  }
  > .actions {
    position: absolute;
    top: 0;
    right: 0;
    height: 2rem;
    transition: opacity 0.1s ease-in-out;
    visibility: hidden;
    opacity: 0;
    // background-color: pink;
    img {
      display: block;
      float: left;
      width: 1.5rem;
      height: 1.5rem;
      margin-top: 0.25rem;
      margin-left: 0.25rem;
      opacity: 0.5;
      border-radius: 4px;
      transition: opacity 0.1s ease-in-out, background-color 0.1s ease-in-out;
      outline: 0;
      // background-color: blue;
      &:focus {
        opacity: 1;
        background-color: ${styleAbstractions.colours.outline};
      }
      &:hover {
        opacity: 1;
      }
    }
  }
  > .tags {
    position: absolute;
    top: 0;
    right: 0;
    height: 2rem;
    span {
      display: block;
      float: left;
      height: 1.5rem;
      line-height: 1.5rem;
      margin-top: 0.25rem;
      padding-left: 0.5rem;
      padding-right: 0.5rem;
      border-radius: 5000px;
      ${styleAbstractions.font}
      font-weight: 800;
      font-size: 70%;
      text-transform: uppercase;
      color: white;
    }
    span + span {
      margin-left: 0.5rem;
    }
  }
`

const actions = {
  'load': (id, handleOnAction) => (<img title='Load' key='action-load' onClick={(e) => handleOnAction(e, id, 'load')} className='action' src={icons.generic.actions.load} alt='' tabIndex='0' />),
  'add': (id, handleOnAction) => (<img title='Add' key='action-add' onClick={(e) => handleOnAction(e, id, 'add')} className='action' src={icons.generic.actions.add} alt='' tabIndex='0' />),
  'move-up': (id, handleOnAction) => (<img title='Move up' key='action-move-up' onClick={(e) => handleOnAction(e, id, 'move-up')} className='action' src={icons.generic.actions.move.up} alt='' tabIndex='0' />),
  'move-down': (id, handleOnAction) => (<img title='Move down' key='action-move-down' onClick={(e) => handleOnAction(e, id, 'move-down')} className='action' src={icons.generic.actions.move.down} alt='' tabIndex='0' />),
  'rename': (id, handleOnAction) => (<img title='Rename' key='action-rename' onClick={(e) => handleOnAction(e, id, 'rename')} className='action' src={icons.generic.actions.edit} alt='' tabIndex='0' />),
  'edit': (id, handleOnAction) => (<img title='Edit' key='action-edit' onClick={(e) => handleOnAction(e, id, 'edit')} className='action' src={icons.generic.actions.edit} alt='' tabIndex='0' />),
  'delete': (id, handleOnAction) => (<img title='Delete' key='action-delete' onClick={(e) => handleOnAction(e, id, 'delete')} className='action' src={icons.generic.actions.delete} alt='' tabIndex='0' />)
}

const getAction = (onAction, id, name) => {
  const foundAction = actions[name]
  if (typeof foundAction === 'function') {
    return foundAction(id, onAction)
  }
  return null
}

const ListItem = ({ id, icon, selected, actions, tags, children, onChoose, onAction, indentation }) => {
  const handleKeypress = (e) => {
    if (e.key === 'Enter') {
      onChoose(id)
    }
  }
  const handleOnAction = (e, id, action) => {
    e.stopPropagation()
    onAction(id, action)
  }
  const imgStyle = {
    marginLeft: `${indentation}px`
  }
  return (
    <StyledListItem onClick={() => onChoose(id)} onTouchEnd={() => onChoose(id)} className={classnames('component--list-item', {'selected': selected})} tabIndex='0' onKeyPress={handleKeypress}>
      <img src={icon} style={imgStyle} alt='' />
      <span>{children}</span>
      <div className='actions'>
        {actions.map(name => getAction(handleOnAction, id, name))}
      </div>
      <div className='tags'>
        {tags.map(tag => (
          <span key={tag.name} style={{backgroundColor: tag.colour}}>{tag.name}</span>
        ))}
      </div>
    </StyledListItem>
  )
}

ListItem.propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  icon: PropTypes.string,
  selected: PropTypes.bool,
  actions: PropTypes.array,
  tags: PropTypes.arrayOf(PropTypes.object),
  onChoose: PropTypes.func,
  onAction: PropTypes.func,
  indentation: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
}

ListItem.defaultProps = {
  icon: '',
  selected: false,
  actions: [],
  tags: [],
  onChoose: () => {},
  onAction: () => {},
  indentation: 4
}

export default ListItem
