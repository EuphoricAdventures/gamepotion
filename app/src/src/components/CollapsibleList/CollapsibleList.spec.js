import React from 'react'
import renderer from 'react-test-renderer'
import { MemoryRouter } from 'react-router-dom'
import 'jest-styled-components'
import CollapsibleList from './CollapsibleList'
import ListItem from '../ListItem/ListItem'

jest.mock('uuid', () => jest.fn().mockReturnValue('id'))

it('renders', () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <CollapsibleList label='stuff'>
          <ListItem id='list-item-1'>List Item 1</ListItem>
          <ListItem id='list-item-2'>List Item 2</ListItem>
        </CollapsibleList>
      </MemoryRouter>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
