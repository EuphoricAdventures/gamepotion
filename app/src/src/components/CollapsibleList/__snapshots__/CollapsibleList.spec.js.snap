// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
.c1 {
  display: block;
  padding: 0.4rem 0.8rem 0.3rem 0.8rem;
  -webkit-transition: background-color 0.3s cubic-bezier(.4,0,.2,1),box-shadow 0.3s cubic-bezier(.4,0,.2,1),opacity 0.2s ease-in-out;
  transition: background-color 0.3s cubic-bezier(.4,0,.2,1),box-shadow 0.3s cubic-bezier(.4,0,.2,1),opacity 0.2s ease-in-out;
  font-family: "Rubik",-apple-system,BlinkMacSystemFont,"Segoe UI",sans-serif;
  font-weight: 400;
  -webkit-text-size-adjust: 100%;
  text-size-adjust: 100%;
  -webkit-font-smoothing: antialiased;
  border-radius: 6px;
  outline: 0;
  box-shadow: 0 3px 1px -2px rgba(0,0,0,.2),0 2px 2px 0 rgba(0,0,0,.14),0 1px 5px 0 rgba(0,0,0,.12);
}

.c1.flavour-strong {
  background-color: #2e3131;
  color: #ffffff;
}

.c1.flavour-strong:not([disabled]):hover {
  background-color: #3C4850;
}

.c1.flavour-strong:focus {
  background-color: #3C4850;
}

.c1.flavour-weak {
  background-color: #eff3f4;
  color: #2e3131;
}

.c1:not([disabled]):hover,
.c1:focus {
  box-shadow: 0 2px 4px -1px rgba(0,0,0,.2),0 4px 5px 0 rgba(0,0,0,.14),0 1px 10px 0 rgba(0,0,0,.12);
}

.c1[disabled] {
  opacity: 0.5;
  cursor: not-allowed;
}

.c1 > img {
  display: block;
  width: 2rem;
  height: 2rem;
  margin: 0 auto 0 auto;
}

.c1 > span {
  display: block;
  height: 1.5rem;
  line-height: 1.5rem;
}

.c1 > img + span {
  margin-top: 0.5rem;
}

.c2 .component--list {
  margin-left: 1.5rem;
}

.c2 .component--list .component--list-item:first-of-type {
  margin-top: 0.25rem;
}

.c2 .component--list + .component--list-item {
  margin-top: 0.25rem;
}

.c2 .component--list-item + .component--list-item {
  margin-top: 0.25rem;
}

.c2 .empty-text {
  padding-top: 2rem;
  padding-bottom: 2rem;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",sans-serif;
  font-weight: 400;
  -webkit-text-size-adjust: 100%;
  text-size-adjust: 100%;
  -webkit-font-smoothing: antialiased;
  text-align: center;
  color: #2e3131;
  opacity: 0.5;
}

.c0 .component--button {
  width: 100%;
  opacity: 0.5;
}

.c0 .component--button + .component--list {
  margin-top: 0.5rem;
}

.c3 {
  position: relative;
  display: block;
  width: calc(100% - 4px);
  height: 2rem;
  list-style-type: none;
  border-width: 2px;
  border-style: solid;
  border-color: transparent;
  border-radius: 4px;
  background-color: transparent;
  -webkit-transition: border-color 0.2s ease-in-out;
  transition: border-color 0.2s ease-in-out;
  outline: 0;
  cursor: default;
}

.c3.selected {
  background-color: #dadfe1;
}

.c3:focus {
  border-color: #a537fd;
}

.c3 > img {
  display: block;
  float: left;
  width: 1.5rem;
  height: 1.5rem;
  margin-top: 0.3rem;
  margin-bottom: 0.3rem;
  margin-left: 0.3rem;
}

.c3 > span {
  display: block;
  float: left;
  height: 1.5rem;
  line-height: 1.5rem;
  margin-top: 0.3rem;
  margin-bottom: 0.3rem;
  margin-left: 0.5rem;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",sans-serif;
  font-weight: 400;
  -webkit-text-size-adjust: 100%;
  text-size-adjust: 100%;
  -webkit-font-smoothing: antialiased;
  color: #2e3131;
}

.c3:hover > .actions {
  visibility: visible;
  opacity: 1;
}

.c3 > .actions {
  position: absolute;
  top: 0;
  right: 0;
  height: 2rem;
  -webkit-transition: opacity 0.1s ease-in-out;
  transition: opacity 0.1s ease-in-out;
  visibility: hidden;
  opacity: 0;
}

.c3 > .actions img {
  display: block;
  float: left;
  width: 1.5rem;
  height: 1.5rem;
  margin-top: 0.25rem;
  margin-left: 0.25rem;
  opacity: 0.5;
  border-radius: 4px;
  -webkit-transition: opacity 0.1s ease-in-out,background-color 0.1s ease-in-out;
  transition: opacity 0.1s ease-in-out,background-color 0.1s ease-in-out;
  outline: 0;
}

.c3 > .actions img:focus {
  opacity: 1;
  background-color: #a537fd;
}

.c3 > .actions img:hover {
  opacity: 1;
}

.c3 > .tags {
  position: absolute;
  top: 0;
  right: 0;
  height: 2rem;
}

.c3 > .tags span {
  display: block;
  float: left;
  height: 1.5rem;
  line-height: 1.5rem;
  margin-top: 0.25rem;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  border-radius: 5000px;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",sans-serif;
  font-weight: 400;
  -webkit-text-size-adjust: 100%;
  text-size-adjust: 100%;
  -webkit-font-smoothing: antialiased;
  font-weight: 800;
  font-size: 70%;
  text-transform: uppercase;
  color: white;
}

.c3 > .tags span + span {
  margin-left: 0.5rem;
}

<ul
  className="component--list c0"
>
  <button
    className="component--button flavour-weak c1"
    disabled={false}
    onClick={[Function]}
    title=""
  >
    <span>
      Hide
       
      stuff
    </span>
  </button>
  <ul
    className="component--list c2"
  >
    <li
      className="component--list-item c3"
      onClick={[Function]}
      onKeyPress={[Function]}
      onTouchEnd={[Function]}
      tabIndex="0"
    >
      <img
        alt=""
        src=""
        style={
          Object {
            "marginLeft": "4px",
          }
        }
      />
      <span>
        List Item 1
      </span>
      <div
        className="actions"
      />
      <div
        className="tags"
      />
    </li>
    <li
      className="component--list-item c3"
      onClick={[Function]}
      onKeyPress={[Function]}
      onTouchEnd={[Function]}
      tabIndex="0"
    >
      <img
        alt=""
        src=""
        style={
          Object {
            "marginLeft": "4px",
          }
        }
      />
      <span>
        List Item 2
      </span>
      <div
        className="actions"
      />
      <div
        className="tags"
      />
    </li>
  </ul>
</ul>
`;
